{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-05-20 14:11:05.557254: I tensorflow/core/platform/cpu_feature_guard.cc:193] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 AVX512F AVX512_VNNI FMA\n",
      "To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.\n",
      "2024-05-20 14:11:05.706513: I tensorflow/core/util/port.cc:104] oneDNN custom operations are on. You may see slightly different numerical results due to floating-point round-off errors from different computation orders. To turn them off, set the environment variable `TF_ENABLE_ONEDNN_OPTS=0`.\n",
      "2024-05-20 14:11:05.710343: W tensorflow/compiler/xla/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libcudart.so.11.0'; dlerror: libcudart.so.11.0: cannot open shared object file: No such file or directory; LD_LIBRARY_PATH: /data/twang15/alleleCount-4.2.1/lib/\n",
      "2024-05-20 14:11:05.710357: I tensorflow/compiler/xla/stream_executor/cuda/cudart_stub.cc:29] Ignore above cudart dlerror if you do not have a GPU set up on your machine.\n",
      "2024-05-20 14:11:06.381051: W tensorflow/compiler/xla/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libnvinfer.so.7'; dlerror: libnvinfer.so.7: cannot open shared object file: No such file or directory; LD_LIBRARY_PATH: /data/twang15/alleleCount-4.2.1/lib/\n",
      "2024-05-20 14:11:06.381114: W tensorflow/compiler/xla/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libnvinfer_plugin.so.7'; dlerror: libnvinfer_plugin.so.7: cannot open shared object file: No such file or directory; LD_LIBRARY_PATH: /data/twang15/alleleCount-4.2.1/lib/\n",
      "2024-05-20 14:11:06.381121: W tensorflow/compiler/tf2tensorrt/utils/py_utils.cc:38] TF-TRT Warning: Cannot dlopen some TensorRT libraries. If you would like to use Nvidia GPU with TensorRT, please make sure the missing libraries mentioned above are installed properly.\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "from scipy.io import mmread\n",
    "import os\n",
    "import matplotlib.pyplot as plt\n",
    "plt.rcParams[\"figure.figsize\"] = (6, 4)\n",
    "\n",
    "from sklearn.metrics import confusion_matrix, ConfusionMatrixDisplay\n",
    "\n",
    "import anndata as ad\n",
    "import scanpy as sc\n",
    "import maxfuse as mf\n",
    "from tqdm import tqdm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "rna_adata = sc.read_h5ad(\"/data/twang15/spatial_protein/CODEX_data/hubmap/scRNA/large_intestine/large_intestine.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>assay_ontology_term_id</th>\n",
       "      <th>donor_id</th>\n",
       "      <th>anatomical_information</th>\n",
       "      <th>n_counts_UMIs</th>\n",
       "      <th>n_genes</th>\n",
       "      <th>cell_ontology_class</th>\n",
       "      <th>free_annotation</th>\n",
       "      <th>manually_annotated</th>\n",
       "      <th>compartment</th>\n",
       "      <th>sex_ontology_term_id</th>\n",
       "      <th>...</th>\n",
       "      <th>tissue_type</th>\n",
       "      <th>cell_type</th>\n",
       "      <th>assay</th>\n",
       "      <th>disease</th>\n",
       "      <th>organism</th>\n",
       "      <th>sex</th>\n",
       "      <th>tissue</th>\n",
       "      <th>self_reported_ethnicity</th>\n",
       "      <th>development_stage</th>\n",
       "      <th>observation_joinid</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACCCAAGCGCCTTG_TSP14_LI_Distal_10X_1_1</th>\n",
       "      <td>EFO:0009922</td>\n",
       "      <td>TSP14</td>\n",
       "      <td>Distal</td>\n",
       "      <td>4572.0</td>\n",
       "      <td>1490</td>\n",
       "      <td>cd4-positive, alpha-beta t cell</td>\n",
       "      <td>cd4-positive alpha-beta T cell</td>\n",
       "      <td>True</td>\n",
       "      <td>immune</td>\n",
       "      <td>PATO:0000384</td>\n",
       "      <td>...</td>\n",
       "      <td>tissue</td>\n",
       "      <td>CD4-positive, alpha-beta T cell</td>\n",
       "      <td>10x 3' v3</td>\n",
       "      <td>normal</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>male</td>\n",
       "      <td>large intestine</td>\n",
       "      <td>European</td>\n",
       "      <td>59-year-old human stage</td>\n",
       "      <td>=?9TSos%=i</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCCACACTTGGGC_TSP14_LI_Distal_10X_1_1</th>\n",
       "      <td>EFO:0009922</td>\n",
       "      <td>TSP14</td>\n",
       "      <td>Distal</td>\n",
       "      <td>12874.0</td>\n",
       "      <td>1466</td>\n",
       "      <td>enterocyte of epithelium of large intestine</td>\n",
       "      <td>immature enterocyte</td>\n",
       "      <td>True</td>\n",
       "      <td>epithelial</td>\n",
       "      <td>PATO:0000384</td>\n",
       "      <td>...</td>\n",
       "      <td>tissue</td>\n",
       "      <td>enterocyte of epithelium of large intestine</td>\n",
       "      <td>10x 3' v3</td>\n",
       "      <td>normal</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>male</td>\n",
       "      <td>large intestine</td>\n",
       "      <td>European</td>\n",
       "      <td>59-year-old human stage</td>\n",
       "      <td>@%;bf?*BFM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCCACATTCAGCA_TSP14_LI_Distal_10X_1_1</th>\n",
       "      <td>EFO:0009922</td>\n",
       "      <td>TSP14</td>\n",
       "      <td>Distal</td>\n",
       "      <td>4400.0</td>\n",
       "      <td>1666</td>\n",
       "      <td>monocyte</td>\n",
       "      <td>monocyte</td>\n",
       "      <td>True</td>\n",
       "      <td>immune</td>\n",
       "      <td>PATO:0000384</td>\n",
       "      <td>...</td>\n",
       "      <td>tissue</td>\n",
       "      <td>monocyte</td>\n",
       "      <td>10x 3' v3</td>\n",
       "      <td>normal</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>male</td>\n",
       "      <td>large intestine</td>\n",
       "      <td>European</td>\n",
       "      <td>59-year-old human stage</td>\n",
       "      <td>*^tPAZZp(j</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCCAGTATCTTCT_TSP14_LI_Distal_10X_1_1</th>\n",
       "      <td>EFO:0009922</td>\n",
       "      <td>TSP14</td>\n",
       "      <td>Distal</td>\n",
       "      <td>13988.0</td>\n",
       "      <td>2319</td>\n",
       "      <td>enterocyte of epithelium of large intestine</td>\n",
       "      <td>immature enterocyte</td>\n",
       "      <td>True</td>\n",
       "      <td>epithelial</td>\n",
       "      <td>PATO:0000384</td>\n",
       "      <td>...</td>\n",
       "      <td>tissue</td>\n",
       "      <td>enterocyte of epithelium of large intestine</td>\n",
       "      <td>10x 3' v3</td>\n",
       "      <td>normal</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>male</td>\n",
       "      <td>large intestine</td>\n",
       "      <td>European</td>\n",
       "      <td>59-year-old human stage</td>\n",
       "      <td>uLV;rz?W@i</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCCAGTGGTATGG_TSP14_LI_Distal_10X_1_1</th>\n",
       "      <td>EFO:0009922</td>\n",
       "      <td>TSP14</td>\n",
       "      <td>Distal</td>\n",
       "      <td>6641.0</td>\n",
       "      <td>1372</td>\n",
       "      <td>enterocyte of epithelium of large intestine</td>\n",
       "      <td>immature enterocyte</td>\n",
       "      <td>True</td>\n",
       "      <td>epithelial</td>\n",
       "      <td>PATO:0000384</td>\n",
       "      <td>...</td>\n",
       "      <td>tissue</td>\n",
       "      <td>enterocyte of epithelium of large intestine</td>\n",
       "      <td>10x 3' v3</td>\n",
       "      <td>normal</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>male</td>\n",
       "      <td>large intestine</td>\n",
       "      <td>European</td>\n",
       "      <td>59-year-old human stage</td>\n",
       "      <td>BbZFljt&gt;80</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TSP2_LI_proximal_SS2_B114618_B134450_Stromal_P21_S381</th>\n",
       "      <td>EFO:0008931</td>\n",
       "      <td>TSP2</td>\n",
       "      <td>proximal</td>\n",
       "      <td>183430.0</td>\n",
       "      <td>1015</td>\n",
       "      <td>goblet cell</td>\n",
       "      <td>goblet cell</td>\n",
       "      <td>True</td>\n",
       "      <td>epithelial</td>\n",
       "      <td>PATO:0000383</td>\n",
       "      <td>...</td>\n",
       "      <td>tissue</td>\n",
       "      <td>large intestine goblet cell</td>\n",
       "      <td>Smart-seq2</td>\n",
       "      <td>normal</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>large intestine</td>\n",
       "      <td>African American or Afro-Caribbean</td>\n",
       "      <td>61-year-old human stage</td>\n",
       "      <td>p6HnaCnt(?</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TSP2_LI_proximal_SS2_B114618_B134450_Stromal_P22_S382</th>\n",
       "      <td>EFO:0008931</td>\n",
       "      <td>TSP2</td>\n",
       "      <td>proximal</td>\n",
       "      <td>61756.0</td>\n",
       "      <td>672</td>\n",
       "      <td>intestinal tuft cell</td>\n",
       "      <td>intestinal tuft cell</td>\n",
       "      <td>True</td>\n",
       "      <td>epithelial</td>\n",
       "      <td>PATO:0000383</td>\n",
       "      <td>...</td>\n",
       "      <td>tissue</td>\n",
       "      <td>intestinal tuft cell</td>\n",
       "      <td>Smart-seq2</td>\n",
       "      <td>normal</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>large intestine</td>\n",
       "      <td>African American or Afro-Caribbean</td>\n",
       "      <td>61-year-old human stage</td>\n",
       "      <td>0+UFL@Uj*#</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TSP2_LI_proximal_SS2_B114618_B134450_Stromal_P23_S383</th>\n",
       "      <td>EFO:0008931</td>\n",
       "      <td>TSP2</td>\n",
       "      <td>proximal</td>\n",
       "      <td>164072.0</td>\n",
       "      <td>1377</td>\n",
       "      <td>goblet cell</td>\n",
       "      <td>goblet cell</td>\n",
       "      <td>True</td>\n",
       "      <td>epithelial</td>\n",
       "      <td>PATO:0000383</td>\n",
       "      <td>...</td>\n",
       "      <td>tissue</td>\n",
       "      <td>large intestine goblet cell</td>\n",
       "      <td>Smart-seq2</td>\n",
       "      <td>normal</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>large intestine</td>\n",
       "      <td>African American or Afro-Caribbean</td>\n",
       "      <td>61-year-old human stage</td>\n",
       "      <td>U7n6*(2t2d</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TSP2_LI_proximal_SS2_B114618_B134450_Stromal_P3_S363</th>\n",
       "      <td>EFO:0008931</td>\n",
       "      <td>TSP2</td>\n",
       "      <td>proximal</td>\n",
       "      <td>462660.0</td>\n",
       "      <td>370</td>\n",
       "      <td>mature enterocyte</td>\n",
       "      <td>mature enterocyte</td>\n",
       "      <td>True</td>\n",
       "      <td>epithelial</td>\n",
       "      <td>PATO:0000383</td>\n",
       "      <td>...</td>\n",
       "      <td>tissue</td>\n",
       "      <td>enterocyte of epithelium of large intestine</td>\n",
       "      <td>Smart-seq2</td>\n",
       "      <td>normal</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>large intestine</td>\n",
       "      <td>African American or Afro-Caribbean</td>\n",
       "      <td>61-year-old human stage</td>\n",
       "      <td>2iWlg3Oi}!</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TSP2_LI_proximal_SS2_B114618_B134450_Stromal_P8_S368</th>\n",
       "      <td>EFO:0008931</td>\n",
       "      <td>TSP2</td>\n",
       "      <td>proximal</td>\n",
       "      <td>172887.0</td>\n",
       "      <td>418</td>\n",
       "      <td>mature enterocyte</td>\n",
       "      <td>mature enterocyte</td>\n",
       "      <td>True</td>\n",
       "      <td>epithelial</td>\n",
       "      <td>PATO:0000383</td>\n",
       "      <td>...</td>\n",
       "      <td>tissue</td>\n",
       "      <td>enterocyte of epithelium of large intestine</td>\n",
       "      <td>Smart-seq2</td>\n",
       "      <td>normal</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>large intestine</td>\n",
       "      <td>African American or Afro-Caribbean</td>\n",
       "      <td>61-year-old human stage</td>\n",
       "      <td>a1z;~1|oiQ</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>13680 rows × 28 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                   assay_ontology_term_id  \\\n",
       "cell_id                                                                     \n",
       "AAACCCAAGCGCCTTG_TSP14_LI_Distal_10X_1_1                      EFO:0009922   \n",
       "AAACCCACACTTGGGC_TSP14_LI_Distal_10X_1_1                      EFO:0009922   \n",
       "AAACCCACATTCAGCA_TSP14_LI_Distal_10X_1_1                      EFO:0009922   \n",
       "AAACCCAGTATCTTCT_TSP14_LI_Distal_10X_1_1                      EFO:0009922   \n",
       "AAACCCAGTGGTATGG_TSP14_LI_Distal_10X_1_1                      EFO:0009922   \n",
       "...                                                                   ...   \n",
       "TSP2_LI_proximal_SS2_B114618_B134450_Stromal_P2...            EFO:0008931   \n",
       "TSP2_LI_proximal_SS2_B114618_B134450_Stromal_P2...            EFO:0008931   \n",
       "TSP2_LI_proximal_SS2_B114618_B134450_Stromal_P2...            EFO:0008931   \n",
       "TSP2_LI_proximal_SS2_B114618_B134450_Stromal_P3...            EFO:0008931   \n",
       "TSP2_LI_proximal_SS2_B114618_B134450_Stromal_P8...            EFO:0008931   \n",
       "\n",
       "                                                   donor_id  \\\n",
       "cell_id                                                       \n",
       "AAACCCAAGCGCCTTG_TSP14_LI_Distal_10X_1_1              TSP14   \n",
       "AAACCCACACTTGGGC_TSP14_LI_Distal_10X_1_1              TSP14   \n",
       "AAACCCACATTCAGCA_TSP14_LI_Distal_10X_1_1              TSP14   \n",
       "AAACCCAGTATCTTCT_TSP14_LI_Distal_10X_1_1              TSP14   \n",
       "AAACCCAGTGGTATGG_TSP14_LI_Distal_10X_1_1              TSP14   \n",
       "...                                                     ...   \n",
       "TSP2_LI_proximal_SS2_B114618_B134450_Stromal_P2...     TSP2   \n",
       "TSP2_LI_proximal_SS2_B114618_B134450_Stromal_P2...     TSP2   \n",
       "TSP2_LI_proximal_SS2_B114618_B134450_Stromal_P2...     TSP2   \n",
       "TSP2_LI_proximal_SS2_B114618_B134450_Stromal_P3...     TSP2   \n",
       "TSP2_LI_proximal_SS2_B114618_B134450_Stromal_P8...     TSP2   \n",
       "\n",
       "                                                   anatomical_information  \\\n",
       "cell_id                                                                     \n",
       "AAACCCAAGCGCCTTG_TSP14_LI_Distal_10X_1_1                           Distal   \n",
       "AAACCCACACTTGGGC_TSP14_LI_Distal_10X_1_1                           Distal   \n",
       "AAACCCACATTCAGCA_TSP14_LI_Distal_10X_1_1                           Distal   \n",
       "AAACCCAGTATCTTCT_TSP14_LI_Distal_10X_1_1                           Distal   \n",
       "AAACCCAGTGGTATGG_TSP14_LI_Distal_10X_1_1                           Distal   \n",
       "...                                                                   ...   \n",
       "TSP2_LI_proximal_SS2_B114618_B134450_Stromal_P2...               proximal   \n",
       "TSP2_LI_proximal_SS2_B114618_B134450_Stromal_P2...               proximal   \n",
       "TSP2_LI_proximal_SS2_B114618_B134450_Stromal_P2...               proximal   \n",
       "TSP2_LI_proximal_SS2_B114618_B134450_Stromal_P3...               proximal   \n",
       "TSP2_LI_proximal_SS2_B114618_B134450_Stromal_P8...               proximal   \n",
       "\n",
       "                                                    n_counts_UMIs  n_genes  \\\n",
       "cell_id                                                                      \n",
       "AAACCCAAGCGCCTTG_TSP14_LI_Distal_10X_1_1                   4572.0     1490   \n",
       "AAACCCACACTTGGGC_TSP14_LI_Distal_10X_1_1                  12874.0     1466   \n",
       "AAACCCACATTCAGCA_TSP14_LI_Distal_10X_1_1                   4400.0     1666   \n",
       "AAACCCAGTATCTTCT_TSP14_LI_Distal_10X_1_1                  13988.0     2319   \n",
       "AAACCCAGTGGTATGG_TSP14_LI_Distal_10X_1_1                   6641.0     1372   \n",
       "...                                                           ...      ...   \n",
       "TSP2_LI_proximal_SS2_B114618_B134450_Stromal_P2...       183430.0     1015   \n",
       "TSP2_LI_proximal_SS2_B114618_B134450_Stromal_P2...        61756.0      672   \n",
       "TSP2_LI_proximal_SS2_B114618_B134450_Stromal_P2...       164072.0     1377   \n",
       "TSP2_LI_proximal_SS2_B114618_B134450_Stromal_P3...       462660.0      370   \n",
       "TSP2_LI_proximal_SS2_B114618_B134450_Stromal_P8...       172887.0      418   \n",
       "\n",
       "                                                                            cell_ontology_class  \\\n",
       "cell_id                                                                                           \n",
       "AAACCCAAGCGCCTTG_TSP14_LI_Distal_10X_1_1                        cd4-positive, alpha-beta t cell   \n",
       "AAACCCACACTTGGGC_TSP14_LI_Distal_10X_1_1            enterocyte of epithelium of large intestine   \n",
       "AAACCCACATTCAGCA_TSP14_LI_Distal_10X_1_1                                               monocyte   \n",
       "AAACCCAGTATCTTCT_TSP14_LI_Distal_10X_1_1            enterocyte of epithelium of large intestine   \n",
       "AAACCCAGTGGTATGG_TSP14_LI_Distal_10X_1_1            enterocyte of epithelium of large intestine   \n",
       "...                                                                                         ...   \n",
       "TSP2_LI_proximal_SS2_B114618_B134450_Stromal_P2...                                  goblet cell   \n",
       "TSP2_LI_proximal_SS2_B114618_B134450_Stromal_P2...                         intestinal tuft cell   \n",
       "TSP2_LI_proximal_SS2_B114618_B134450_Stromal_P2...                                  goblet cell   \n",
       "TSP2_LI_proximal_SS2_B114618_B134450_Stromal_P3...                            mature enterocyte   \n",
       "TSP2_LI_proximal_SS2_B114618_B134450_Stromal_P8...                            mature enterocyte   \n",
       "\n",
       "                                                                   free_annotation  \\\n",
       "cell_id                                                                              \n",
       "AAACCCAAGCGCCTTG_TSP14_LI_Distal_10X_1_1            cd4-positive alpha-beta T cell   \n",
       "AAACCCACACTTGGGC_TSP14_LI_Distal_10X_1_1                       immature enterocyte   \n",
       "AAACCCACATTCAGCA_TSP14_LI_Distal_10X_1_1                                  monocyte   \n",
       "AAACCCAGTATCTTCT_TSP14_LI_Distal_10X_1_1                       immature enterocyte   \n",
       "AAACCCAGTGGTATGG_TSP14_LI_Distal_10X_1_1                       immature enterocyte   \n",
       "...                                                                            ...   \n",
       "TSP2_LI_proximal_SS2_B114618_B134450_Stromal_P2...                     goblet cell   \n",
       "TSP2_LI_proximal_SS2_B114618_B134450_Stromal_P2...            intestinal tuft cell   \n",
       "TSP2_LI_proximal_SS2_B114618_B134450_Stromal_P2...                     goblet cell   \n",
       "TSP2_LI_proximal_SS2_B114618_B134450_Stromal_P3...               mature enterocyte   \n",
       "TSP2_LI_proximal_SS2_B114618_B134450_Stromal_P8...               mature enterocyte   \n",
       "\n",
       "                                                    manually_annotated  \\\n",
       "cell_id                                                                  \n",
       "AAACCCAAGCGCCTTG_TSP14_LI_Distal_10X_1_1                          True   \n",
       "AAACCCACACTTGGGC_TSP14_LI_Distal_10X_1_1                          True   \n",
       "AAACCCACATTCAGCA_TSP14_LI_Distal_10X_1_1                          True   \n",
       "AAACCCAGTATCTTCT_TSP14_LI_Distal_10X_1_1                          True   \n",
       "AAACCCAGTGGTATGG_TSP14_LI_Distal_10X_1_1                          True   \n",
       "...                                                                ...   \n",
       "TSP2_LI_proximal_SS2_B114618_B134450_Stromal_P2...                True   \n",
       "TSP2_LI_proximal_SS2_B114618_B134450_Stromal_P2...                True   \n",
       "TSP2_LI_proximal_SS2_B114618_B134450_Stromal_P2...                True   \n",
       "TSP2_LI_proximal_SS2_B114618_B134450_Stromal_P3...                True   \n",
       "TSP2_LI_proximal_SS2_B114618_B134450_Stromal_P8...                True   \n",
       "\n",
       "                                                   compartment  \\\n",
       "cell_id                                                          \n",
       "AAACCCAAGCGCCTTG_TSP14_LI_Distal_10X_1_1                immune   \n",
       "AAACCCACACTTGGGC_TSP14_LI_Distal_10X_1_1            epithelial   \n",
       "AAACCCACATTCAGCA_TSP14_LI_Distal_10X_1_1                immune   \n",
       "AAACCCAGTATCTTCT_TSP14_LI_Distal_10X_1_1            epithelial   \n",
       "AAACCCAGTGGTATGG_TSP14_LI_Distal_10X_1_1            epithelial   \n",
       "...                                                        ...   \n",
       "TSP2_LI_proximal_SS2_B114618_B134450_Stromal_P2...  epithelial   \n",
       "TSP2_LI_proximal_SS2_B114618_B134450_Stromal_P2...  epithelial   \n",
       "TSP2_LI_proximal_SS2_B114618_B134450_Stromal_P2...  epithelial   \n",
       "TSP2_LI_proximal_SS2_B114618_B134450_Stromal_P3...  epithelial   \n",
       "TSP2_LI_proximal_SS2_B114618_B134450_Stromal_P8...  epithelial   \n",
       "\n",
       "                                                   sex_ontology_term_id  ...  \\\n",
       "cell_id                                                                  ...   \n",
       "AAACCCAAGCGCCTTG_TSP14_LI_Distal_10X_1_1                   PATO:0000384  ...   \n",
       "AAACCCACACTTGGGC_TSP14_LI_Distal_10X_1_1                   PATO:0000384  ...   \n",
       "AAACCCACATTCAGCA_TSP14_LI_Distal_10X_1_1                   PATO:0000384  ...   \n",
       "AAACCCAGTATCTTCT_TSP14_LI_Distal_10X_1_1                   PATO:0000384  ...   \n",
       "AAACCCAGTGGTATGG_TSP14_LI_Distal_10X_1_1                   PATO:0000384  ...   \n",
       "...                                                                 ...  ...   \n",
       "TSP2_LI_proximal_SS2_B114618_B134450_Stromal_P2...         PATO:0000383  ...   \n",
       "TSP2_LI_proximal_SS2_B114618_B134450_Stromal_P2...         PATO:0000383  ...   \n",
       "TSP2_LI_proximal_SS2_B114618_B134450_Stromal_P2...         PATO:0000383  ...   \n",
       "TSP2_LI_proximal_SS2_B114618_B134450_Stromal_P3...         PATO:0000383  ...   \n",
       "TSP2_LI_proximal_SS2_B114618_B134450_Stromal_P8...         PATO:0000383  ...   \n",
       "\n",
       "                                                   tissue_type  \\\n",
       "cell_id                                                          \n",
       "AAACCCAAGCGCCTTG_TSP14_LI_Distal_10X_1_1                tissue   \n",
       "AAACCCACACTTGGGC_TSP14_LI_Distal_10X_1_1                tissue   \n",
       "AAACCCACATTCAGCA_TSP14_LI_Distal_10X_1_1                tissue   \n",
       "AAACCCAGTATCTTCT_TSP14_LI_Distal_10X_1_1                tissue   \n",
       "AAACCCAGTGGTATGG_TSP14_LI_Distal_10X_1_1                tissue   \n",
       "...                                                        ...   \n",
       "TSP2_LI_proximal_SS2_B114618_B134450_Stromal_P2...      tissue   \n",
       "TSP2_LI_proximal_SS2_B114618_B134450_Stromal_P2...      tissue   \n",
       "TSP2_LI_proximal_SS2_B114618_B134450_Stromal_P2...      tissue   \n",
       "TSP2_LI_proximal_SS2_B114618_B134450_Stromal_P3...      tissue   \n",
       "TSP2_LI_proximal_SS2_B114618_B134450_Stromal_P8...      tissue   \n",
       "\n",
       "                                                                                      cell_type  \\\n",
       "cell_id                                                                                           \n",
       "AAACCCAAGCGCCTTG_TSP14_LI_Distal_10X_1_1                        CD4-positive, alpha-beta T cell   \n",
       "AAACCCACACTTGGGC_TSP14_LI_Distal_10X_1_1            enterocyte of epithelium of large intestine   \n",
       "AAACCCACATTCAGCA_TSP14_LI_Distal_10X_1_1                                               monocyte   \n",
       "AAACCCAGTATCTTCT_TSP14_LI_Distal_10X_1_1            enterocyte of epithelium of large intestine   \n",
       "AAACCCAGTGGTATGG_TSP14_LI_Distal_10X_1_1            enterocyte of epithelium of large intestine   \n",
       "...                                                                                         ...   \n",
       "TSP2_LI_proximal_SS2_B114618_B134450_Stromal_P2...                  large intestine goblet cell   \n",
       "TSP2_LI_proximal_SS2_B114618_B134450_Stromal_P2...                         intestinal tuft cell   \n",
       "TSP2_LI_proximal_SS2_B114618_B134450_Stromal_P2...                  large intestine goblet cell   \n",
       "TSP2_LI_proximal_SS2_B114618_B134450_Stromal_P3...  enterocyte of epithelium of large intestine   \n",
       "TSP2_LI_proximal_SS2_B114618_B134450_Stromal_P8...  enterocyte of epithelium of large intestine   \n",
       "\n",
       "                                                         assay disease  \\\n",
       "cell_id                                                                  \n",
       "AAACCCAAGCGCCTTG_TSP14_LI_Distal_10X_1_1             10x 3' v3  normal   \n",
       "AAACCCACACTTGGGC_TSP14_LI_Distal_10X_1_1             10x 3' v3  normal   \n",
       "AAACCCACATTCAGCA_TSP14_LI_Distal_10X_1_1             10x 3' v3  normal   \n",
       "AAACCCAGTATCTTCT_TSP14_LI_Distal_10X_1_1             10x 3' v3  normal   \n",
       "AAACCCAGTGGTATGG_TSP14_LI_Distal_10X_1_1             10x 3' v3  normal   \n",
       "...                                                        ...     ...   \n",
       "TSP2_LI_proximal_SS2_B114618_B134450_Stromal_P2...  Smart-seq2  normal   \n",
       "TSP2_LI_proximal_SS2_B114618_B134450_Stromal_P2...  Smart-seq2  normal   \n",
       "TSP2_LI_proximal_SS2_B114618_B134450_Stromal_P2...  Smart-seq2  normal   \n",
       "TSP2_LI_proximal_SS2_B114618_B134450_Stromal_P3...  Smart-seq2  normal   \n",
       "TSP2_LI_proximal_SS2_B114618_B134450_Stromal_P8...  Smart-seq2  normal   \n",
       "\n",
       "                                                        organism     sex  \\\n",
       "cell_id                                                                    \n",
       "AAACCCAAGCGCCTTG_TSP14_LI_Distal_10X_1_1            Homo sapiens    male   \n",
       "AAACCCACACTTGGGC_TSP14_LI_Distal_10X_1_1            Homo sapiens    male   \n",
       "AAACCCACATTCAGCA_TSP14_LI_Distal_10X_1_1            Homo sapiens    male   \n",
       "AAACCCAGTATCTTCT_TSP14_LI_Distal_10X_1_1            Homo sapiens    male   \n",
       "AAACCCAGTGGTATGG_TSP14_LI_Distal_10X_1_1            Homo sapiens    male   \n",
       "...                                                          ...     ...   \n",
       "TSP2_LI_proximal_SS2_B114618_B134450_Stromal_P2...  Homo sapiens  female   \n",
       "TSP2_LI_proximal_SS2_B114618_B134450_Stromal_P2...  Homo sapiens  female   \n",
       "TSP2_LI_proximal_SS2_B114618_B134450_Stromal_P2...  Homo sapiens  female   \n",
       "TSP2_LI_proximal_SS2_B114618_B134450_Stromal_P3...  Homo sapiens  female   \n",
       "TSP2_LI_proximal_SS2_B114618_B134450_Stromal_P8...  Homo sapiens  female   \n",
       "\n",
       "                                                             tissue  \\\n",
       "cell_id                                                               \n",
       "AAACCCAAGCGCCTTG_TSP14_LI_Distal_10X_1_1            large intestine   \n",
       "AAACCCACACTTGGGC_TSP14_LI_Distal_10X_1_1            large intestine   \n",
       "AAACCCACATTCAGCA_TSP14_LI_Distal_10X_1_1            large intestine   \n",
       "AAACCCAGTATCTTCT_TSP14_LI_Distal_10X_1_1            large intestine   \n",
       "AAACCCAGTGGTATGG_TSP14_LI_Distal_10X_1_1            large intestine   \n",
       "...                                                             ...   \n",
       "TSP2_LI_proximal_SS2_B114618_B134450_Stromal_P2...  large intestine   \n",
       "TSP2_LI_proximal_SS2_B114618_B134450_Stromal_P2...  large intestine   \n",
       "TSP2_LI_proximal_SS2_B114618_B134450_Stromal_P2...  large intestine   \n",
       "TSP2_LI_proximal_SS2_B114618_B134450_Stromal_P3...  large intestine   \n",
       "TSP2_LI_proximal_SS2_B114618_B134450_Stromal_P8...  large intestine   \n",
       "\n",
       "                                                               self_reported_ethnicity  \\\n",
       "cell_id                                                                                  \n",
       "AAACCCAAGCGCCTTG_TSP14_LI_Distal_10X_1_1                                      European   \n",
       "AAACCCACACTTGGGC_TSP14_LI_Distal_10X_1_1                                      European   \n",
       "AAACCCACATTCAGCA_TSP14_LI_Distal_10X_1_1                                      European   \n",
       "AAACCCAGTATCTTCT_TSP14_LI_Distal_10X_1_1                                      European   \n",
       "AAACCCAGTGGTATGG_TSP14_LI_Distal_10X_1_1                                      European   \n",
       "...                                                                                ...   \n",
       "TSP2_LI_proximal_SS2_B114618_B134450_Stromal_P2...  African American or Afro-Caribbean   \n",
       "TSP2_LI_proximal_SS2_B114618_B134450_Stromal_P2...  African American or Afro-Caribbean   \n",
       "TSP2_LI_proximal_SS2_B114618_B134450_Stromal_P2...  African American or Afro-Caribbean   \n",
       "TSP2_LI_proximal_SS2_B114618_B134450_Stromal_P3...  African American or Afro-Caribbean   \n",
       "TSP2_LI_proximal_SS2_B114618_B134450_Stromal_P8...  African American or Afro-Caribbean   \n",
       "\n",
       "                                                          development_stage  \\\n",
       "cell_id                                                                       \n",
       "AAACCCAAGCGCCTTG_TSP14_LI_Distal_10X_1_1            59-year-old human stage   \n",
       "AAACCCACACTTGGGC_TSP14_LI_Distal_10X_1_1            59-year-old human stage   \n",
       "AAACCCACATTCAGCA_TSP14_LI_Distal_10X_1_1            59-year-old human stage   \n",
       "AAACCCAGTATCTTCT_TSP14_LI_Distal_10X_1_1            59-year-old human stage   \n",
       "AAACCCAGTGGTATGG_TSP14_LI_Distal_10X_1_1            59-year-old human stage   \n",
       "...                                                                     ...   \n",
       "TSP2_LI_proximal_SS2_B114618_B134450_Stromal_P2...  61-year-old human stage   \n",
       "TSP2_LI_proximal_SS2_B114618_B134450_Stromal_P2...  61-year-old human stage   \n",
       "TSP2_LI_proximal_SS2_B114618_B134450_Stromal_P2...  61-year-old human stage   \n",
       "TSP2_LI_proximal_SS2_B114618_B134450_Stromal_P3...  61-year-old human stage   \n",
       "TSP2_LI_proximal_SS2_B114618_B134450_Stromal_P8...  61-year-old human stage   \n",
       "\n",
       "                                                   observation_joinid  \n",
       "cell_id                                                                \n",
       "AAACCCAAGCGCCTTG_TSP14_LI_Distal_10X_1_1                   =?9TSos%=i  \n",
       "AAACCCACACTTGGGC_TSP14_LI_Distal_10X_1_1                   @%;bf?*BFM  \n",
       "AAACCCACATTCAGCA_TSP14_LI_Distal_10X_1_1                   *^tPAZZp(j  \n",
       "AAACCCAGTATCTTCT_TSP14_LI_Distal_10X_1_1                   uLV;rz?W@i  \n",
       "AAACCCAGTGGTATGG_TSP14_LI_Distal_10X_1_1                   BbZFljt>80  \n",
       "...                                                               ...  \n",
       "TSP2_LI_proximal_SS2_B114618_B134450_Stromal_P2...         p6HnaCnt(?  \n",
       "TSP2_LI_proximal_SS2_B114618_B134450_Stromal_P2...         0+UFL@Uj*#  \n",
       "TSP2_LI_proximal_SS2_B114618_B134450_Stromal_P2...         U7n6*(2t2d  \n",
       "TSP2_LI_proximal_SS2_B114618_B134450_Stromal_P3...         2iWlg3Oi}!  \n",
       "TSP2_LI_proximal_SS2_B114618_B134450_Stromal_P8...         a1z;~1|oiQ  \n",
       "\n",
       "[13680 rows x 28 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rna_adata.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/data/twang15/mamba/envs/imc/lib/python3.8/site-packages/anndata/_core/anndata.py:1838: UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"obs\")\n"
     ]
    }
   ],
   "source": [
    "protein_adata = sc.read_h5ad(\"/data/twang15/spatial_protein/CODEX_data/hubmap/anndata/large_intestine/combined_protein.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/data/twang15/mamba/envs/imc/lib/python3.8/site-packages/anndata/_core/anndata.py:887: UserWarning: \n",
      "AnnData expects .var.index to contain strings, but got values like:\n",
      "    ['DDX11L1', 'WASH7P', 'MIR6859-1', 'MIR1302-2HG', 'MIR1302-2']\n",
      "\n",
      "    Inferred to be: categorical\n",
      "\n",
      "  names = self._prep_dim_index(names, \"var\")\n"
     ]
    }
   ],
   "source": [
    "rna_adata.var_names = rna_adata.var['feature_name']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/data/twang15/mamba/envs/imc/lib/python3.8/site-packages/anndata/_core/anndata.py:121: ImplicitModificationWarning: Transforming to str index.\n",
      "  warnings.warn(\"Transforming to str index.\", ImplicitModificationWarning)\n"
     ]
    }
   ],
   "source": [
    "correspondence = pd.read_csv(\"/data/twang15/maxfuse/docs/protein_gene_conversion.csv\")\n",
    "rna_protein_correspondence = []\n",
    "\n",
    "for i in range(correspondence.shape[0]):\n",
    "    curr_protein_name, curr_rna_names = correspondence.iloc[i]\n",
    "    if curr_protein_name not in protein_adata.var_names:\n",
    "        continue\n",
    "    if curr_rna_names.find('Ignore') != -1: # some correspondence ignored eg. protein isoform to one gene\n",
    "        continue\n",
    "    curr_rna_names = curr_rna_names.split('/') # eg. one protein to multiple genes\n",
    "    for r in curr_rna_names:\n",
    "        if r in rna_adata.var_names:\n",
    "            rna_protein_correspondence.append([r, curr_protein_name])\n",
    "            \n",
    "\n",
    "rna_protein_correspondence = np.array(rna_protein_correspondence)\n",
    "_, unique_indices = np.unique(rna_protein_correspondence[:, 1], return_index=True)\n",
    "rna_protein_correspondence = rna_protein_correspondence[sorted(unique_indices)]\n",
    "_, unique_indices = np.unique(rna_protein_correspondence[:, 0], return_index=True)\n",
    "rna_protein_correspondence = rna_protein_correspondence[sorted(unique_indices)]\n",
    "\n",
    "\n",
    "protein_adata.obs_names_make_unique\n",
    "rna_adata.obs_names_make_unique\n",
    "protein_adata.var_names_make_unique\n",
    "rna_adata.var_names_make_unique\n",
    "mask1 = rna_adata.var_names.isin(rna_protein_correspondence[:, 0])\n",
    "mask2 = protein_adata.var_names.isin(rna_protein_correspondence[:, 1])\n",
    "\n",
    "rna_shared = rna_adata[:, mask1].copy()\n",
    "# rna_shared = rna_adata.loc[:, rna_protein_correspondence[:, 0]].copy()\n",
    "protein_shared = protein_adata[:, mask2].copy() \n",
    "\n",
    "mask = (\n",
    "    (rna_shared.X.toarray().std(axis=0) > 1e-5) \n",
    "    & (protein_shared.X.std(axis=0) > 1e-5)\n",
    ")\n",
    "rna_shared = rna_shared[:, mask].copy()\n",
    "protein_shared = protein_shared[:, mask].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/data/twang15/mamba/envs/imc/lib/python3.8/site-packages/scanpy/preprocessing/_normalization.py:196: UserWarning: Some cells have zero counts\n",
      "  warn(UserWarning('Some cells have zero counts'))\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(13680, 47778)\n",
      "(1027537, 24)\n",
      "(13680, 22)\n",
      "(1027537, 22)\n"
     ]
    }
   ],
   "source": [
    "# row sum for RNA\n",
    "rna_counts = rna_shared.X.toarray().sum(axis=1)\n",
    "# row sum for protein\n",
    "protein_counts = protein_shared.X.sum(axis=1)\n",
    "# take median of each and then take mean\n",
    "target_sum = (np.median(rna_counts.copy()) + np.median(protein_counts.copy())) / 2\n",
    "\n",
    "\n",
    "# process rna_shared\n",
    "sc.pp.normalize_total(rna_shared, target_sum=target_sum)\n",
    "sc.pp.log1p(rna_shared)\n",
    "sc.pp.scale(rna_shared)\n",
    "\n",
    "# sc.pp.neighbors(rna_shared, n_neighbors=15)\n",
    "# sc.tl.umap(rna_shared)\n",
    "# sc.pl.umap(rna_shared, color=['celltype.l1','celltype.l2'])\n",
    "\n",
    "rna_shared = rna_shared.X.copy()\n",
    "\n",
    "# process protein_shared\n",
    "sc.pp.normalize_total(protein_shared, target_sum=target_sum)\n",
    "sc.pp.log1p(protein_shared)\n",
    "sc.pp.scale(protein_shared)\n",
    "\n",
    "protein_shared = protein_shared.X.copy()\n",
    "\n",
    "# process all RNA features\n",
    "sc.pp.normalize_total(rna_adata)\n",
    "sc.pp.log1p(rna_adata)\n",
    "sc.pp.highly_variable_genes(rna_adata)\n",
    "# only retain highly variable genes\n",
    "# rna_adata = rna_adata[:, rna_adata.var.highly_variable].copy()\n",
    "sc.pp.scale(rna_adata)\n",
    "\n",
    "\n",
    "# process all protein features\n",
    "sc.pp.normalize_total(protein_adata)\n",
    "sc.pp.log1p(protein_adata)\n",
    "sc.pp.scale(protein_adata)\n",
    "\n",
    "\n",
    "# make sure no feature is static\n",
    "rna_active = rna_adata.X\n",
    "protein_active = protein_adata.X\n",
    "rna_active = rna_active[:, rna_active.std(axis=0) > 1e-5]\n",
    "protein_active = protein_active[:, protein_active.std(axis=0) > 1e-5]\n",
    "# inspect shape of the four matrices\n",
    "print(rna_active.shape)\n",
    "print(protein_active.shape)\n",
    "print(rna_shared.shape)\n",
    "print(protein_shared.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The first data is split into 1 batches, average batch size is 13680, and max batch size is 13680.\n",
      "The second data is split into 25 batches, average batch size is 41101, and max batch size is 41113.\n",
      "Batch to batch correspondence is:\n",
      "  ['0<->0', '0<->1', '0<->2', '0<->3', '0<->4', '0<->5', '0<->6', '0<->7', '0<->8', '0<->9', '0<->10', '0<->11', '0<->12', '0<->13', '0<->14', '0<->15', '0<->16', '0<->17', '0<->18', '0<->19', '0<->20', '0<->21', '0<->22', '0<->23', '0<->24'].\n"
     ]
    }
   ],
   "source": [
    "fusor = mf.model.Fusor(\n",
    "    shared_arr1=rna_shared,\n",
    "    shared_arr2=protein_shared,\n",
    "    active_arr1=rna_active,\n",
    "    active_arr2=protein_active,\n",
    "    labels1=None,\n",
    "    labels2=None\n",
    ")\n",
    "fusor.split_into_batches(\n",
    "    max_outward_size=10000,\n",
    "    matching_ratio=4,\n",
    "    metacell_size=2,\n",
    "    verbose=True\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Aggregating cells in arr1 into metacells of average size 2...\n",
      "Constructing neighborhood graphs for cells in arr1...\n",
      "Now at batch 0...\n",
      "Graph construction finished!\n",
      "Clustering into metacells...\n",
      "Now at batch 0...\n",
      "Metacell clustering finished!\n",
      "Constructing neighborhood graphs for cells in arr1...\n",
      "Now at batch 0...\n",
      "Graph construction finished!\n",
      "Clustering the graphs for cells in arr1...\n",
      "Now at batch 0...\n",
      "Graph clustering finished!\n",
      "Constructing neighborhood graphs for cells in arr2...\n",
      "Now at batch 0...\n",
      "Now at batch 1...\n",
      "Now at batch 2...\n",
      "Now at batch 3...\n",
      "Now at batch 4...\n",
      "Now at batch 5...\n",
      "Now at batch 6...\n",
      "Now at batch 7...\n",
      "Now at batch 8...\n",
      "Now at batch 9...\n",
      "Now at batch 10...\n",
      "Now at batch 11...\n",
      "Now at batch 12...\n",
      "Now at batch 13...\n",
      "Now at batch 14...\n",
      "Now at batch 15...\n",
      "Now at batch 16...\n",
      "Now at batch 17...\n",
      "Now at batch 18...\n",
      "Now at batch 19...\n",
      "Now at batch 20...\n",
      "Now at batch 21...\n",
      "Now at batch 22...\n",
      "Now at batch 23...\n",
      "Now at batch 24...\n",
      "Graph construction finished!\n",
      "Clustering the graphs for cells in arr2...\n",
      "Now at batch 0...\n",
      "Now at batch 1...\n",
      "Now at batch 2...\n",
      "Now at batch 3...\n",
      "Now at batch 4...\n",
      "Now at batch 5...\n",
      "Now at batch 6...\n",
      "Now at batch 7...\n",
      "Now at batch 8...\n",
      "Now at batch 9...\n",
      "Now at batch 10...\n",
      "Now at batch 11...\n",
      "Now at batch 12...\n",
      "Now at batch 13...\n",
      "Now at batch 14...\n",
      "Now at batch 15...\n",
      "Now at batch 16...\n",
      "Now at batch 17...\n",
      "Now at batch 18...\n",
      "Now at batch 19...\n",
      "Now at batch 20...\n",
      "Now at batch 21...\n",
      "Now at batch 22...\n",
      "Now at batch 23...\n",
      "Now at batch 24...\n",
      "Graph clustering finished!\n"
     ]
    }
   ],
   "source": [
    "fusor.construct_graphs(\n",
    "    n_neighbors1=15,\n",
    "    n_neighbors2=15,\n",
    "    svd_components1=10,\n",
    "    svd_components2=10,\n",
    "    resolution1=2,\n",
    "    resolution2=2,\n",
    "    randomized_svd=False,\n",
    "    svd_runs=1,\n",
    "    resolution_tol=0.1,\n",
    "    leiden_runs=1,\n",
    "    leiden_seed=None,\n",
    "    verbose=True\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Now at batch 0<->0...\n",
      "Now at batch 0<->1...\n",
      "Now at batch 0<->2...\n",
      "Now at batch 0<->3...\n",
      "Now at batch 0<->4...\n",
      "Now at batch 0<->5...\n",
      "Now at batch 0<->6...\n",
      "Now at batch 0<->7...\n",
      "Now at batch 0<->8...\n",
      "Now at batch 0<->9...\n",
      "Now at batch 0<->10...\n",
      "Now at batch 0<->11...\n",
      "Now at batch 0<->12...\n",
      "Now at batch 0<->13...\n",
      "Now at batch 0<->14...\n",
      "Now at batch 0<->15...\n",
      "Now at batch 0<->16...\n",
      "Now at batch 0<->17...\n",
      "Now at batch 0<->18...\n",
      "Now at batch 0<->19...\n",
      "Now at batch 0<->20...\n",
      "Now at batch 0<->21...\n",
      "Now at batch 0<->22...\n",
      "Now at batch 0<->23...\n",
      "Now at batch 0<->24...\n",
      "Done!\n"
     ]
    }
   ],
   "source": [
    "fusor.find_initial_pivots(\n",
    "    wt1=0.7, wt2=0.7,\n",
    "    svd_components1=15, svd_components2=10\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Now at batch 0<->0...\n",
      "Now at batch 0<->1...\n",
      "Now at batch 0<->2...\n",
      "Now at batch 0<->3...\n",
      "Now at batch 0<->4...\n",
      "Now at batch 0<->5...\n",
      "Now at batch 0<->6...\n",
      "Now at batch 0<->7...\n",
      "Now at batch 0<->8...\n",
      "Now at batch 0<->9...\n",
      "Now at batch 0<->10...\n",
      "Now at batch 0<->11...\n",
      "Now at batch 0<->12...\n",
      "Now at batch 0<->13...\n",
      "Now at batch 0<->14...\n",
      "Now at batch 0<->15...\n",
      "Now at batch 0<->16...\n",
      "Now at batch 0<->17...\n",
      "Now at batch 0<->18...\n",
      "Now at batch 0<->19...\n",
      "Now at batch 0<->20...\n",
      "Now at batch 0<->21...\n",
      "Now at batch 0<->22...\n",
      "Now at batch 0<->23...\n",
      "Now at batch 0<->24...\n",
      "Done!\n",
      "Begin filtering...\n",
      "Now at batch 0<->0...\n",
      "Now at batch 0<->1...\n",
      "Now at batch 0<->2...\n",
      "Now at batch 0<->3...\n",
      "Now at batch 0<->4...\n",
      "Now at batch 0<->5...\n",
      "Now at batch 0<->6...\n",
      "Now at batch 0<->7...\n",
      "Now at batch 0<->8...\n",
      "Now at batch 0<->9...\n",
      "Now at batch 0<->10...\n",
      "Now at batch 0<->11...\n",
      "Now at batch 0<->12...\n",
      "Now at batch 0<->13...\n",
      "Now at batch 0<->14...\n",
      "Now at batch 0<->15...\n",
      "Now at batch 0<->16...\n",
      "Now at batch 0<->17...\n",
      "Now at batch 0<->18...\n",
      "Now at batch 0<->19...\n",
      "Now at batch 0<->20...\n",
      "Now at batch 0<->21...\n",
      "Now at batch 0<->22...\n",
      "Now at batch 0<->23...\n",
      "Now at batch 0<->24...\n",
      "162950/171525 pairs of matched cells remain after the filtering.\n",
      "Fitting CCA on pivots...\n",
      "Scoring matched pairs...\n",
      "13331/13680 cells in arr1 are selected as pivots.\n",
      "162950/1027537 cells in arr2 are selected as pivots.\n",
      "Done!\n",
      "Now at batch 0<->0...\n",
      "Now at batch 0<->1...\n",
      "Now at batch 0<->2...\n",
      "Now at batch 0<->3...\n",
      "Now at batch 0<->4...\n",
      "Now at batch 0<->5...\n",
      "Now at batch 0<->6...\n",
      "Now at batch 0<->7...\n",
      "Now at batch 0<->8...\n",
      "Now at batch 0<->9...\n",
      "Now at batch 0<->10...\n",
      "Now at batch 0<->11...\n",
      "Now at batch 0<->12...\n",
      "Now at batch 0<->13...\n",
      "Now at batch 0<->14...\n",
      "Now at batch 0<->15...\n",
      "Now at batch 0<->16...\n",
      "Now at batch 0<->17...\n",
      "Now at batch 0<->18...\n",
      "Now at batch 0<->19...\n",
      "Now at batch 0<->20...\n",
      "Now at batch 0<->21...\n",
      "Now at batch 0<->22...\n",
      "Now at batch 0<->23...\n",
      "Now at batch 0<->24...\n",
      "Done!\n",
      "Begin filtering...\n",
      "Now at batch 0<->0...\n",
      "Now at batch 0<->1...\n",
      "Now at batch 0<->2...\n",
      "Now at batch 0<->3...\n",
      "Now at batch 0<->4...\n",
      "Now at batch 0<->5...\n",
      "Now at batch 0<->6...\n",
      "Now at batch 0<->7...\n",
      "Now at batch 0<->8...\n",
      "Now at batch 0<->9...\n",
      "Now at batch 0<->10...\n",
      "Now at batch 0<->11...\n",
      "Now at batch 0<->12...\n",
      "Now at batch 0<->13...\n",
      "Now at batch 0<->14...\n",
      "Now at batch 0<->15...\n",
      "Now at batch 0<->16...\n",
      "Now at batch 0<->17...\n",
      "Now at batch 0<->18...\n",
      "Now at batch 0<->19...\n",
      "Now at batch 0<->20...\n",
      "Now at batch 0<->21...\n",
      "Now at batch 0<->22...\n",
      "Now at batch 0<->23...\n",
      "Now at batch 0<->24...\n",
      "611208/873162 pairs of matched cells remain after the filtering.\n",
      "Scoring matched pairs...\n",
      "Done!\n"
     ]
    }
   ],
   "source": [
    "fusor.refine_pivots(\n",
    "    wt1=0.3, wt2=0.3,\n",
    "    svd_components1=40, svd_components2=None,\n",
    "    cca_components=20,\n",
    "    n_iters=1,\n",
    "    randomized_svd=False,\n",
    "    svd_runs=1,\n",
    "    verbose=True\n",
    ")\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "fusor.filter_bad_matches(target='pivot', filter_prop=0.05)\n",
    "\n",
    "\n",
    "\n",
    "pivot_matching = fusor.get_matching(target='pivot')\n",
    "\n",
    "\n",
    "\n",
    "fusor.propagate(\n",
    "    svd_components1=40,\n",
    "    svd_components2=None,\n",
    "    wt1=0.7,\n",
    "    wt2=0.7,\n",
    ")\n",
    "fusor.filter_bad_matches(\n",
    "    target='propagated',\n",
    "    filter_prop=0.3\n",
    ")\n",
    "full_matching = fusor.get_matching(order=(2, 1), target='full_data')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "matching = pd.DataFrame(list(zip(full_matching[0], full_matching[1], full_matching[2])), columns = ['mod1_indx', 'mod2_indx', 'score'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "matching.to_csv('/data/twang15/spatial_protein/CODEX_data/hubmap/integrated/large_intestine/matching.tsv', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/data/twang15/mamba/envs/imc/lib/python3.8/site-packages/anndata/_core/anndata.py:1838: UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"obs\")\n",
      "/data/twang15/mamba/envs/imc/lib/python3.8/site-packages/anndata/_core/anndata.py:121: ImplicitModificationWarning: Transforming to str index.\n",
      "  warnings.warn(\"Transforming to str index.\", ImplicitModificationWarning)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                                   assay_ontology_term_id  \\\n",
      "cell_id                                                                     \n",
      "AGTCACACAAGCAGGT_TSP14_LI_Distal_10X_1_1                      EFO:0009922   \n",
      "CCTAACCAGAAAGTCT_TSP14_LI_Distal_10X_1_1                      EFO:0009922   \n",
      "TTCTCTCCAAATGGCG_TSP14_LI_Distal_10X_1_1                      EFO:0009922   \n",
      "CAACAGTGTAGTTAGA_TSP14_LI_Proximal_10X_1_1                    EFO:0009922   \n",
      "TTACTGTTCGTTCTAT_TSP14_LI_Distal_10X_1_1                      EFO:0009922   \n",
      "...                                                                   ...   \n",
      "ACCCTTGCATACAGAA_TSP2_LI_distal_10X_1_1_Sigmoid               EFO:0009922   \n",
      "CGTAGTAGTTGGTAGG_TSP2_LI_distal_10X_1_1_Sigmoid               EFO:0009922   \n",
      "GGGAAGTCAATTTCTC_TSP2_LI_proximal_10X_1_1_Ascen...            EFO:0009922   \n",
      "TSP2_LI_proximal_SS2_B114586_B134454_Epithelial...            EFO:0008931   \n",
      "GTAGAAACACGAAAGC_TSP2_LI_distal_10X_1_1_Sigmoid               EFO:0009922   \n",
      "\n",
      "                                                   donor_id  \\\n",
      "cell_id                                                       \n",
      "AGTCACACAAGCAGGT_TSP14_LI_Distal_10X_1_1              TSP14   \n",
      "CCTAACCAGAAAGTCT_TSP14_LI_Distal_10X_1_1              TSP14   \n",
      "TTCTCTCCAAATGGCG_TSP14_LI_Distal_10X_1_1              TSP14   \n",
      "CAACAGTGTAGTTAGA_TSP14_LI_Proximal_10X_1_1            TSP14   \n",
      "TTACTGTTCGTTCTAT_TSP14_LI_Distal_10X_1_1              TSP14   \n",
      "...                                                     ...   \n",
      "ACCCTTGCATACAGAA_TSP2_LI_distal_10X_1_1_Sigmoid        TSP2   \n",
      "CGTAGTAGTTGGTAGG_TSP2_LI_distal_10X_1_1_Sigmoid        TSP2   \n",
      "GGGAAGTCAATTTCTC_TSP2_LI_proximal_10X_1_1_Ascen...     TSP2   \n",
      "TSP2_LI_proximal_SS2_B114586_B134454_Epithelial...     TSP2   \n",
      "GTAGAAACACGAAAGC_TSP2_LI_distal_10X_1_1_Sigmoid        TSP2   \n",
      "\n",
      "                                                   anatomical_information  \\\n",
      "cell_id                                                                     \n",
      "AGTCACACAAGCAGGT_TSP14_LI_Distal_10X_1_1                           Distal   \n",
      "CCTAACCAGAAAGTCT_TSP14_LI_Distal_10X_1_1                           Distal   \n",
      "TTCTCTCCAAATGGCG_TSP14_LI_Distal_10X_1_1                           Distal   \n",
      "CAACAGTGTAGTTAGA_TSP14_LI_Proximal_10X_1_1                       Proximal   \n",
      "TTACTGTTCGTTCTAT_TSP14_LI_Distal_10X_1_1                           Distal   \n",
      "...                                                                   ...   \n",
      "ACCCTTGCATACAGAA_TSP2_LI_distal_10X_1_1_Sigmoid                    distal   \n",
      "CGTAGTAGTTGGTAGG_TSP2_LI_distal_10X_1_1_Sigmoid                    distal   \n",
      "GGGAAGTCAATTTCTC_TSP2_LI_proximal_10X_1_1_Ascen...               proximal   \n",
      "TSP2_LI_proximal_SS2_B114586_B134454_Epithelial...                    nan   \n",
      "GTAGAAACACGAAAGC_TSP2_LI_distal_10X_1_1_Sigmoid                    distal   \n",
      "\n",
      "                                                    n_counts_UMIs  n_genes  \\\n",
      "cell_id                                                                      \n",
      "AGTCACACAAGCAGGT_TSP14_LI_Distal_10X_1_1                   4016.0     1233   \n",
      "CCTAACCAGAAAGTCT_TSP14_LI_Distal_10X_1_1                   5910.0     1597   \n",
      "TTCTCTCCAAATGGCG_TSP14_LI_Distal_10X_1_1                   3939.0     1508   \n",
      "CAACAGTGTAGTTAGA_TSP14_LI_Proximal_10X_1_1                 3749.0     1319   \n",
      "TTACTGTTCGTTCTAT_TSP14_LI_Distal_10X_1_1                  30032.0     4420   \n",
      "...                                                           ...      ...   \n",
      "ACCCTTGCATACAGAA_TSP2_LI_distal_10X_1_1_Sigmoid            5930.0     1920   \n",
      "CGTAGTAGTTGGTAGG_TSP2_LI_distal_10X_1_1_Sigmoid           16515.0     3712   \n",
      "GGGAAGTCAATTTCTC_TSP2_LI_proximal_10X_1_1_Ascen...        28284.0     5199   \n",
      "TSP2_LI_proximal_SS2_B114586_B134454_Epithelial...       443476.0     4271   \n",
      "GTAGAAACACGAAAGC_TSP2_LI_distal_10X_1_1_Sigmoid           19806.0     4439   \n",
      "\n",
      "                                                                             cell_ontology_class  \\\n",
      "cell_id                                                                                            \n",
      "AGTCACACAAGCAGGT_TSP14_LI_Distal_10X_1_1             enterocyte of epithelium of large intestine   \n",
      "CCTAACCAGAAAGTCT_TSP14_LI_Distal_10X_1_1                             large intestine goblet cell   \n",
      "TTCTCTCCAAATGGCG_TSP14_LI_Distal_10X_1_1                                                monocyte   \n",
      "CAACAGTGTAGTTAGA_TSP14_LI_Proximal_10X_1_1                       cd4-positive, alpha-beta t cell   \n",
      "TTACTGTTCGTTCTAT_TSP14_LI_Distal_10X_1_1            paneth cell of epithelium of large intestine   \n",
      "...                                                                                          ...   \n",
      "ACCCTTGCATACAGAA_TSP2_LI_distal_10X_1_1_Sigmoid                              immature enterocyte   \n",
      "CGTAGTAGTTGGTAGG_TSP2_LI_distal_10X_1_1_Sigmoid                              immature enterocyte   \n",
      "GGGAAGTCAATTTCTC_TSP2_LI_proximal_10X_1_1_Ascen...                           immature enterocyte   \n",
      "TSP2_LI_proximal_SS2_B114586_B134454_Epithelial...                    intestinal crypt stem cell   \n",
      "GTAGAAACACGAAAGC_TSP2_LI_distal_10X_1_1_Sigmoid                              immature enterocyte   \n",
      "\n",
      "                                                                   free_annotation  \\\n",
      "cell_id                                                                              \n",
      "AGTCACACAAGCAGGT_TSP14_LI_Distal_10X_1_1                       immature enterocyte   \n",
      "CCTAACCAGAAAGTCT_TSP14_LI_Distal_10X_1_1                               goblet cell   \n",
      "TTCTCTCCAAATGGCG_TSP14_LI_Distal_10X_1_1                                  monocyte   \n",
      "CAACAGTGTAGTTAGA_TSP14_LI_Proximal_10X_1_1          cd4-positive alpha-beta T cell   \n",
      "TTACTGTTCGTTCTAT_TSP14_LI_Distal_10X_1_1                          paneth-like cell   \n",
      "...                                                                            ...   \n",
      "ACCCTTGCATACAGAA_TSP2_LI_distal_10X_1_1_Sigmoid                immature enterocyte   \n",
      "CGTAGTAGTTGGTAGG_TSP2_LI_distal_10X_1_1_Sigmoid                immature enterocyte   \n",
      "GGGAAGTCAATTTCTC_TSP2_LI_proximal_10X_1_1_Ascen...             immature enterocyte   \n",
      "TSP2_LI_proximal_SS2_B114586_B134454_Epithelial...      intestinal crypt stem cell   \n",
      "GTAGAAACACGAAAGC_TSP2_LI_distal_10X_1_1_Sigmoid                immature enterocyte   \n",
      "\n",
      "                                                    manually_annotated  \\\n",
      "cell_id                                                                  \n",
      "AGTCACACAAGCAGGT_TSP14_LI_Distal_10X_1_1                          True   \n",
      "CCTAACCAGAAAGTCT_TSP14_LI_Distal_10X_1_1                          True   \n",
      "TTCTCTCCAAATGGCG_TSP14_LI_Distal_10X_1_1                          True   \n",
      "CAACAGTGTAGTTAGA_TSP14_LI_Proximal_10X_1_1                        True   \n",
      "TTACTGTTCGTTCTAT_TSP14_LI_Distal_10X_1_1                          True   \n",
      "...                                                                ...   \n",
      "ACCCTTGCATACAGAA_TSP2_LI_distal_10X_1_1_Sigmoid                   True   \n",
      "CGTAGTAGTTGGTAGG_TSP2_LI_distal_10X_1_1_Sigmoid                   True   \n",
      "GGGAAGTCAATTTCTC_TSP2_LI_proximal_10X_1_1_Ascen...                True   \n",
      "TSP2_LI_proximal_SS2_B114586_B134454_Epithelial...                True   \n",
      "GTAGAAACACGAAAGC_TSP2_LI_distal_10X_1_1_Sigmoid                   True   \n",
      "\n",
      "                                                   compartment  \\\n",
      "cell_id                                                          \n",
      "AGTCACACAAGCAGGT_TSP14_LI_Distal_10X_1_1            epithelial   \n",
      "CCTAACCAGAAAGTCT_TSP14_LI_Distal_10X_1_1            epithelial   \n",
      "TTCTCTCCAAATGGCG_TSP14_LI_Distal_10X_1_1                immune   \n",
      "CAACAGTGTAGTTAGA_TSP14_LI_Proximal_10X_1_1              immune   \n",
      "TTACTGTTCGTTCTAT_TSP14_LI_Distal_10X_1_1            epithelial   \n",
      "...                                                        ...   \n",
      "ACCCTTGCATACAGAA_TSP2_LI_distal_10X_1_1_Sigmoid     epithelial   \n",
      "CGTAGTAGTTGGTAGG_TSP2_LI_distal_10X_1_1_Sigmoid     epithelial   \n",
      "GGGAAGTCAATTTCTC_TSP2_LI_proximal_10X_1_1_Ascen...  epithelial   \n",
      "TSP2_LI_proximal_SS2_B114586_B134454_Epithelial...  epithelial   \n",
      "GTAGAAACACGAAAGC_TSP2_LI_distal_10X_1_1_Sigmoid     epithelial   \n",
      "\n",
      "                                                   sex_ontology_term_id  ...  \\\n",
      "cell_id                                                                  ...   \n",
      "AGTCACACAAGCAGGT_TSP14_LI_Distal_10X_1_1                   PATO:0000384  ...   \n",
      "CCTAACCAGAAAGTCT_TSP14_LI_Distal_10X_1_1                   PATO:0000384  ...   \n",
      "TTCTCTCCAAATGGCG_TSP14_LI_Distal_10X_1_1                   PATO:0000384  ...   \n",
      "CAACAGTGTAGTTAGA_TSP14_LI_Proximal_10X_1_1                 PATO:0000384  ...   \n",
      "TTACTGTTCGTTCTAT_TSP14_LI_Distal_10X_1_1                   PATO:0000384  ...   \n",
      "...                                                                 ...  ...   \n",
      "ACCCTTGCATACAGAA_TSP2_LI_distal_10X_1_1_Sigmoid            PATO:0000383  ...   \n",
      "CGTAGTAGTTGGTAGG_TSP2_LI_distal_10X_1_1_Sigmoid            PATO:0000383  ...   \n",
      "GGGAAGTCAATTTCTC_TSP2_LI_proximal_10X_1_1_Ascen...         PATO:0000383  ...   \n",
      "TSP2_LI_proximal_SS2_B114586_B134454_Epithelial...         PATO:0000383  ...   \n",
      "GTAGAAACACGAAAGC_TSP2_LI_distal_10X_1_1_Sigmoid            PATO:0000383  ...   \n",
      "\n",
      "                                                   Nucleus Boundaries K-Means [Total] Expression  \\\n",
      "cell_id                                                                                            \n",
      "AGTCACACAAGCAGGT_TSP14_LI_Distal_10X_1_1                                                       1   \n",
      "CCTAACCAGAAAGTCT_TSP14_LI_Distal_10X_1_1                                                       1   \n",
      "TTCTCTCCAAATGGCG_TSP14_LI_Distal_10X_1_1                                                       1   \n",
      "CAACAGTGTAGTTAGA_TSP14_LI_Proximal_10X_1_1                                                     1   \n",
      "TTACTGTTCGTTCTAT_TSP14_LI_Distal_10X_1_1                                                       1   \n",
      "...                                                                                          ...   \n",
      "ACCCTTGCATACAGAA_TSP2_LI_distal_10X_1_1_Sigmoid                                                1   \n",
      "CGTAGTAGTTGGTAGG_TSP2_LI_distal_10X_1_1_Sigmoid                                                1   \n",
      "GGGAAGTCAATTTCTC_TSP2_LI_proximal_10X_1_1_Ascen...                                             1   \n",
      "TSP2_LI_proximal_SS2_B114586_B134454_Epithelial...                                             1   \n",
      "GTAGAAACACGAAAGC_TSP2_LI_distal_10X_1_1_Sigmoid                                                1   \n",
      "\n",
      "                                                    Nucleus Boundaries K-Means [Mean-All-SubRegions] Expression  \\\n",
      "cell_id                                                                                                           \n",
      "AGTCACACAAGCAGGT_TSP14_LI_Distal_10X_1_1                                                            1             \n",
      "CCTAACCAGAAAGTCT_TSP14_LI_Distal_10X_1_1                                                            1             \n",
      "TTCTCTCCAAATGGCG_TSP14_LI_Distal_10X_1_1                                                            1             \n",
      "CAACAGTGTAGTTAGA_TSP14_LI_Proximal_10X_1_1                                                          1             \n",
      "TTACTGTTCGTTCTAT_TSP14_LI_Distal_10X_1_1                                                            1             \n",
      "...                                                                                               ...             \n",
      "ACCCTTGCATACAGAA_TSP2_LI_distal_10X_1_1_Sigmoid                                                     1             \n",
      "CGTAGTAGTTGGTAGG_TSP2_LI_distal_10X_1_1_Sigmoid                                                     1             \n",
      "GGGAAGTCAATTTCTC_TSP2_LI_proximal_10X_1_1_Ascen...                                                  1             \n",
      "TSP2_LI_proximal_SS2_B114586_B134454_Epithelial...                                                  1             \n",
      "GTAGAAACACGAAAGC_TSP2_LI_distal_10X_1_1_Sigmoid                                                     1             \n",
      "\n",
      "                                                   Nucleus Boundaries K-Means [Shape-Vectors]  \\\n",
      "cell_id                                                                                         \n",
      "AGTCACACAAGCAGGT_TSP14_LI_Distal_10X_1_1                                                    2   \n",
      "CCTAACCAGAAAGTCT_TSP14_LI_Distal_10X_1_1                                                    1   \n",
      "TTCTCTCCAAATGGCG_TSP14_LI_Distal_10X_1_1                                                    1   \n",
      "CAACAGTGTAGTTAGA_TSP14_LI_Proximal_10X_1_1                                                  1   \n",
      "TTACTGTTCGTTCTAT_TSP14_LI_Distal_10X_1_1                                                    1   \n",
      "...                                                                                       ...   \n",
      "ACCCTTGCATACAGAA_TSP2_LI_distal_10X_1_1_Sigmoid                                             3   \n",
      "CGTAGTAGTTGGTAGG_TSP2_LI_distal_10X_1_1_Sigmoid                                             2   \n",
      "GGGAAGTCAATTTCTC_TSP2_LI_proximal_10X_1_1_Ascen...                                          3   \n",
      "TSP2_LI_proximal_SS2_B114586_B134454_Epithelial...                                          1   \n",
      "GTAGAAACACGAAAGC_TSP2_LI_distal_10X_1_1_Sigmoid                                             2   \n",
      "\n",
      "                                                   Nucleus Boundaries K-Means [Texture]  \\\n",
      "cell_id                                                                                   \n",
      "AGTCACACAAGCAGGT_TSP14_LI_Distal_10X_1_1                                              1   \n",
      "CCTAACCAGAAAGTCT_TSP14_LI_Distal_10X_1_1                                              1   \n",
      "TTCTCTCCAAATGGCG_TSP14_LI_Distal_10X_1_1                                              1   \n",
      "CAACAGTGTAGTTAGA_TSP14_LI_Proximal_10X_1_1                                            1   \n",
      "TTACTGTTCGTTCTAT_TSP14_LI_Distal_10X_1_1                                              1   \n",
      "...                                                                                 ...   \n",
      "ACCCTTGCATACAGAA_TSP2_LI_distal_10X_1_1_Sigmoid                                       1   \n",
      "CGTAGTAGTTGGTAGG_TSP2_LI_distal_10X_1_1_Sigmoid                                       1   \n",
      "GGGAAGTCAATTTCTC_TSP2_LI_proximal_10X_1_1_Ascen...                                    1   \n",
      "TSP2_LI_proximal_SS2_B114586_B134454_Epithelial...                                    1   \n",
      "GTAGAAACACGAAAGC_TSP2_LI_distal_10X_1_1_Sigmoid                                       1   \n",
      "\n",
      "                                                   Nucleus Boundaries K-Means [tSNE_All_Features]  \\\n",
      "cell_id                                                                                             \n",
      "AGTCACACAAGCAGGT_TSP14_LI_Distal_10X_1_1                                                        2   \n",
      "CCTAACCAGAAAGTCT_TSP14_LI_Distal_10X_1_1                                                        1   \n",
      "TTCTCTCCAAATGGCG_TSP14_LI_Distal_10X_1_1                                                        1   \n",
      "CAACAGTGTAGTTAGA_TSP14_LI_Proximal_10X_1_1                                                      2   \n",
      "TTACTGTTCGTTCTAT_TSP14_LI_Distal_10X_1_1                                                        1   \n",
      "...                                                                                           ...   \n",
      "ACCCTTGCATACAGAA_TSP2_LI_distal_10X_1_1_Sigmoid                                                 1   \n",
      "CGTAGTAGTTGGTAGG_TSP2_LI_distal_10X_1_1_Sigmoid                                                 3   \n",
      "GGGAAGTCAATTTCTC_TSP2_LI_proximal_10X_1_1_Ascen...                                              1   \n",
      "TSP2_LI_proximal_SS2_B114586_B134454_Epithelial...                                              2   \n",
      "GTAGAAACACGAAAGC_TSP2_LI_distal_10X_1_1_Sigmoid                                                 1   \n",
      "\n",
      "                                                   Nucleus Boundaries K-Means [Shape-Vectors Normalized]  \\\n",
      "cell_id                                                                                                    \n",
      "AGTCACACAAGCAGGT_TSP14_LI_Distal_10X_1_1                                                            1      \n",
      "CCTAACCAGAAAGTCT_TSP14_LI_Distal_10X_1_1                                                            1      \n",
      "TTCTCTCCAAATGGCG_TSP14_LI_Distal_10X_1_1                                                            1      \n",
      "CAACAGTGTAGTTAGA_TSP14_LI_Proximal_10X_1_1                                                          1      \n",
      "TTACTGTTCGTTCTAT_TSP14_LI_Distal_10X_1_1                                                            1      \n",
      "...                                                                                               ...      \n",
      "ACCCTTGCATACAGAA_TSP2_LI_distal_10X_1_1_Sigmoid                                                     2      \n",
      "CGTAGTAGTTGGTAGG_TSP2_LI_distal_10X_1_1_Sigmoid                                                     2      \n",
      "GGGAAGTCAATTTCTC_TSP2_LI_proximal_10X_1_1_Ascen...                                                  2      \n",
      "TSP2_LI_proximal_SS2_B114586_B134454_Epithelial...                                                  1      \n",
      "GTAGAAACACGAAAGC_TSP2_LI_distal_10X_1_1_Sigmoid                                                     2      \n",
      "\n",
      "                                                   Nucleus Boundaries K-Means [UMAP_All_Features]  \\\n",
      "cell_id                                                                                             \n",
      "AGTCACACAAGCAGGT_TSP14_LI_Distal_10X_1_1                                                        1   \n",
      "CCTAACCAGAAAGTCT_TSP14_LI_Distal_10X_1_1                                                        1   \n",
      "TTCTCTCCAAATGGCG_TSP14_LI_Distal_10X_1_1                                                        2   \n",
      "CAACAGTGTAGTTAGA_TSP14_LI_Proximal_10X_1_1                                                      3   \n",
      "TTACTGTTCGTTCTAT_TSP14_LI_Distal_10X_1_1                                                        2   \n",
      "...                                                                                           ...   \n",
      "ACCCTTGCATACAGAA_TSP2_LI_distal_10X_1_1_Sigmoid                                                 5   \n",
      "CGTAGTAGTTGGTAGG_TSP2_LI_distal_10X_1_1_Sigmoid                                                 5   \n",
      "GGGAAGTCAATTTCTC_TSP2_LI_proximal_10X_1_1_Ascen...                                              5   \n",
      "TSP2_LI_proximal_SS2_B114586_B134454_Epithelial...                                              5   \n",
      "GTAGAAACACGAAAGC_TSP2_LI_distal_10X_1_1_Sigmoid                                                 5   \n",
      "\n",
      "                                                                    Tissue  \\\n",
      "cell_id                                                                      \n",
      "AGTCACACAAGCAGGT_TSP14_LI_Distal_10X_1_1            HBM942.WHFC.488_reg001   \n",
      "CCTAACCAGAAAGTCT_TSP14_LI_Distal_10X_1_1            HBM942.WHFC.488_reg001   \n",
      "TTCTCTCCAAATGGCG_TSP14_LI_Distal_10X_1_1            HBM942.WHFC.488_reg001   \n",
      "CAACAGTGTAGTTAGA_TSP14_LI_Proximal_10X_1_1          HBM942.WHFC.488_reg001   \n",
      "TTACTGTTCGTTCTAT_TSP14_LI_Distal_10X_1_1            HBM942.WHFC.488_reg001   \n",
      "...                                                                    ...   \n",
      "ACCCTTGCATACAGAA_TSP2_LI_distal_10X_1_1_Sigmoid     HBM886.NTZN.682_reg001   \n",
      "CGTAGTAGTTGGTAGG_TSP2_LI_distal_10X_1_1_Sigmoid     HBM886.NTZN.682_reg001   \n",
      "GGGAAGTCAATTTCTC_TSP2_LI_proximal_10X_1_1_Ascen...  HBM886.NTZN.682_reg001   \n",
      "TSP2_LI_proximal_SS2_B114586_B134454_Epithelial...  HBM886.NTZN.682_reg001   \n",
      "GTAGAAACACGAAAGC_TSP2_LI_distal_10X_1_1_Sigmoid     HBM886.NTZN.682_reg001   \n",
      "\n",
      "                                                              x            y  \n",
      "cell_id                                                                       \n",
      "AGTCACACAAGCAGGT_TSP14_LI_Distal_10X_1_1            8575.330484    21.282526  \n",
      "CCTAACCAGAAAGTCT_TSP14_LI_Distal_10X_1_1            8307.690507    21.265930  \n",
      "TTCTCTCCAAATGGCG_TSP14_LI_Distal_10X_1_1             818.589068    23.652506  \n",
      "CAACAGTGTAGTTAGA_TSP14_LI_Proximal_10X_1_1          8778.504958    23.791762  \n",
      "TTACTGTTCGTTCTAT_TSP14_LI_Distal_10X_1_1            2949.433611    24.351080  \n",
      "...                                                         ...          ...  \n",
      "ACCCTTGCATACAGAA_TSP2_LI_distal_10X_1_1_Sigmoid     5027.124533  9437.892359  \n",
      "CGTAGTAGTTGGTAGG_TSP2_LI_distal_10X_1_1_Sigmoid     4746.916131  9447.244851  \n",
      "GGGAAGTCAATTTCTC_TSP2_LI_proximal_10X_1_1_Ascen...  4822.981121  9461.257227  \n",
      "TSP2_LI_proximal_SS2_B114586_B134454_Epithelial...  4758.975464  9466.056782  \n",
      "GTAGAAACACGAAAGC_TSP2_LI_distal_10X_1_1_Sigmoid     4607.450714  9466.634448  \n",
      "\n",
      "[772326 rows x 67 columns]\n",
      "                 feature_type  highly_variable         means  dispersions  \\\n",
      "feature_name                                                                \n",
      "DDX11L1       Gene Expression            False  6.398244e-05     0.835044   \n",
      "WASH7P        Gene Expression            False  2.274395e-03     2.442280   \n",
      "MIR6859-1     Gene Expression            False  6.175251e-05     1.295335   \n",
      "MIR1302-2HG   Gene Expression            False  1.372886e-04     2.656352   \n",
      "MIR1302-2     Gene Expression            False  1.000000e-12          NaN   \n",
      "...                       ...              ...           ...          ...   \n",
      "MT-ND6        Gene Expression            False  9.634841e-01     2.466404   \n",
      "MT-TE         Gene Expression            False  1.600667e-01     1.603787   \n",
      "MT-CYB        Gene Expression            False  4.367693e+00     4.765751   \n",
      "MT-TT         Gene Expression            False  6.573967e-02     0.624316   \n",
      "MT-TP         Gene Expression             True  4.385102e-01     2.705558   \n",
      "\n",
      "              dispersions_norm      mean       std    ensembl_version  \\\n",
      "feature_name                                                            \n",
      "DDX11L1              -0.573947  0.000039  0.005574  ENSG00000223972.5   \n",
      "WASH7P                0.533203  0.001080  0.031731  ENSG00000227232.5   \n",
      "MIR6859-1            -0.256874  0.000033  0.005634  ENSG00000278267.1   \n",
      "MIR1302-2HG           0.680668  0.000048  0.008041  ENSG00000243485.5   \n",
      "MIR1302-2             0.000000  0.000000  1.000000  ENSG00000284332.1   \n",
      "...                        ...       ...       ...                ...   \n",
      "MT-ND6                0.154140  0.590065  0.741395  ENSG00000198695.2   \n",
      "MT-TE                -0.044396  0.083929  0.301820  ENSG00000210194.1   \n",
      "MT-CYB               -0.499747  3.874830  1.104192  ENSG00000198727.2   \n",
      "MT-TT                -0.719108  0.040580  0.186848  ENSG00000210195.2   \n",
      "MT-TP                 1.115482  0.204806  0.501265  ENSG00000210196.2   \n",
      "\n",
      "              feature_is_filtered feature_name feature_reference  \\\n",
      "feature_name                                                       \n",
      "DDX11L1                     False      DDX11L1    NCBITaxon:9606   \n",
      "WASH7P                      False       WASH7P    NCBITaxon:9606   \n",
      "MIR6859-1                   False    MIR6859-1    NCBITaxon:9606   \n",
      "MIR1302-2HG                 False  MIR1302-2HG    NCBITaxon:9606   \n",
      "MIR1302-2                   False    MIR1302-2    NCBITaxon:9606   \n",
      "...                           ...          ...               ...   \n",
      "MT-ND6                      False       MT-ND6    NCBITaxon:9606   \n",
      "MT-TE                       False        MT-TE    NCBITaxon:9606   \n",
      "MT-CYB                      False       MT-CYB    NCBITaxon:9606   \n",
      "MT-TT                       False        MT-TT    NCBITaxon:9606   \n",
      "MT-TP                       False        MT-TP    NCBITaxon:9606   \n",
      "\n",
      "             feature_biotype feature_length  \n",
      "feature_name                                 \n",
      "DDX11L1                 gene            632  \n",
      "WASH7P                  gene           1351  \n",
      "MIR6859-1               gene             68  \n",
      "MIR1302-2HG             gene           1021  \n",
      "MIR1302-2               gene            138  \n",
      "...                      ...            ...  \n",
      "MT-ND6                  gene            525  \n",
      "MT-TE                   gene             69  \n",
      "MT-CYB                  gene           1141  \n",
      "MT-TT                   gene             66  \n",
      "MT-TP                   gene             68  \n",
      "\n",
      "[58482 rows x 13 columns]\n"
     ]
    }
   ],
   "source": [
    "protein_inte = protein_adata[full_matching[1]].copy()\n",
    "\n",
    "rna_adata_for_integrated = rna_adata[full_matching[0]]\n",
    "\n",
    "protein_integrated = rna_adata_for_integrated.copy()\n",
    "protein_inte.obs_names = protein_integrated.obs_names\n",
    "# protein_integrated.obs = protein_inte.obs\n",
    "protein_integrated.obs = pd.concat([protein_integrated.obs, protein_inte.obs], axis=1)\n",
    "\n",
    "print(protein_integrated.obs)\n",
    "print(protein_integrated.var)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "protein_integrated.write_h5ad('/data/twang15/spatial_protein/CODEX_data/hubmap/integrated/large_intestine/combined_protein_integrated.h5ad')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "full_matching = pd.read_csv('/data/twang15/spatial_protein/CODEX_data/hubmap/integrated/large_intestine/matching.tsv', sep='\\t', index_col='Unnamed: 0')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mod1_indx</th>\n",
       "      <th>mod2_indx</th>\n",
       "      <th>score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1282</td>\n",
       "      <td>8</td>\n",
       "      <td>0.560078</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2695</td>\n",
       "      <td>9</td>\n",
       "      <td>0.258774</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>7345</td>\n",
       "      <td>13</td>\n",
       "      <td>0.680557</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>8574</td>\n",
       "      <td>15</td>\n",
       "      <td>0.512414</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>7168</td>\n",
       "      <td>17</td>\n",
       "      <td>0.387523</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>772321</th>\n",
       "      <td>11521</td>\n",
       "      <td>1027532</td>\n",
       "      <td>0.751603</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>772322</th>\n",
       "      <td>11846</td>\n",
       "      <td>1027533</td>\n",
       "      <td>0.301009</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>772323</th>\n",
       "      <td>12944</td>\n",
       "      <td>1027534</td>\n",
       "      <td>0.406181</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>772324</th>\n",
       "      <td>13529</td>\n",
       "      <td>1027535</td>\n",
       "      <td>0.601993</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>772325</th>\n",
       "      <td>12088</td>\n",
       "      <td>1027536</td>\n",
       "      <td>0.614682</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>772326 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        mod1_indx  mod2_indx     score\n",
       "0            1282          8  0.560078\n",
       "1            2695          9  0.258774\n",
       "2            7345         13  0.680557\n",
       "3            8574         15  0.512414\n",
       "4            7168         17  0.387523\n",
       "...           ...        ...       ...\n",
       "772321      11521    1027532  0.751603\n",
       "772322      11846    1027533  0.301009\n",
       "772323      12944    1027534  0.406181\n",
       "772324      13529    1027535  0.601993\n",
       "772325      12088    1027536  0.614682\n",
       "\n",
       "[772326 rows x 3 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "matching"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "imc",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.19"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
